interfaces/prompting/requestrules/requestrules.go

```diff
@@ -0,0 +1,877 @@
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	changeOccurred := false
+	defer func() {
+		if changeOccurred {
+			rdb.save()
+			rdb.notifyRule(user, id, nil)
+		}
+	}()
+	origRule, err := rdb.ruleWithIDForUser(user, id)
+	if err != nil {
+		return nil, err
+	}
+	if constraints == nil {
+		constraints = origRule.Constraints
+	}
+	if outcome == prompting.OutcomeUnset {
+		outcome = origRule.Outcome
+	}
+	if lifespan == prompting.LifespanUnset {
+		lifespan = origRule.Lifespan
+	}
+	if err = rdb.removeRuleFromTree(origRule); err != nil {
+		// If error occurs, rule is fully removed from tree
+		changeOccurred = true
+		return nil, err
+	}
+	newRule, err := rdb.PopulateNewRule(user, origRule.Snap, origRule.Interface, constraints, outcome, lifespan, duration)
+	if err != nil {
+		rdb.addRuleToTree(origRule) // ignore any new error, should not occur
+		// origRule was successfully removed before, so it should now be able
+		// to be successfully re-added without error, and all is unchanged.
+		changeOccurred = false
+		return nil, err
+	}
+	newRule.ID = origRule.ID
+	err, conflicts, conflictingPermission := rdb.addRuleToTree(newRule)
+	if err != nil {
+		rdb.addRuleToTree(origRule) // ignore any new error
+		// origRule was successfully removed before, so it should now be able
+		// to be successfully re-added without error, and all is unchanged.
+		changeOccurred = false
+		return nil, fmt.Errorf("%w: conflicts: %+v, Permission: '%s'", err, conflicts, conflictingPermission)
+	}
+	rdb.removeRuleWithID(origRule.ID) // no error can occur, we just confirmed the rule exists
+	if err := rdb.addRule(newRule); err != nil {
+		// Should not occur
+		return nil, fmt.Errorf("internal error: %v", err)
+	}
+	changeOccurred = true
+	return newRule, nil
+}
+
+// Rules returns all rules which apply to the given user.
+func (rdb *RuleDB) Rules(user uint32) []*Rule {
+	rdb.mutex.Lock()
+	defer rdb.mutex.Unlock()
+	ruleFilter := func(rule *Rule) bool {
+		return rule.User == user
+	}
+	return rdb.rulesInternal(ruleFilter)
+}
+
+// rulesInternal returns all rules matching the given filter.
+//
+// TODO: store rules separately per user, snap, and interface, so actions which
+// look up or delete all rules for a given user/snap/interface are much faster.
+// This is safe, since rules must each apply to exactly one user, snap and
+// interface, but may apply to multiple permissions.
+func (rdb *RuleDB) rulesInternal(ruleFilter func(rule *Rule) bool) []*Rule {
+	rules := make([]*Rule, 0)
+	currTime := time.Now()
+	needToSave := false
+	defer func() {
+		if needToSave {
+			rdb.save()
+		}
+	}()
+	for _, rule := range rdb.rules {
+		expired, err := rule.Expired(currTime)
+		if err != nil {
+			// Issue with expiration, this should not occur
+		}
+		if expired {
+			needToSave = true
+			rdb.removeRuleFromTree(rule)
+			rdb.removeRuleWithID(rule.ID)
+			data := map[string]string{"removed": "expired"}
+			rdb.notifyRule(rule.User, rule.ID, data)
+			continue
+		}
```

Comment:
Yes, this is a better approach. In the past, checking whether a rule was expired was slow, since timestamps were stored as strings which had to be converted to times, so the thought was to remove them as early as possible to reduce checks on known expired rules in the future. But now we're storing them as `time.Time`, so expiration is as simple as an int comparison.

Furthermore, this was always racy, since if we're e.g. returning a list of unexpired rules over the API, and are careful to prune expired rules first, more might expire as we're marshalling/sending, so there's no guarantee that some of the expired rules aren't expired. So there's no need to prune expired rules at all, since clients must know how to deal with expired rules. And removing rules from the tree is slow, so let's avoid doing it even when saving rules.

The one exception is if we're adding/patching a rule and it conflicts with an expired rule, that expired rule should be removed immediately. But we're already planning to mutate the rule DB, so this is fine.